src/sync/wait_group.rs

```diff
@@ -0,0 +1,127 @@
+use std::fmt;
+use std::process;
+use std::sync::Arc;
+
+use parking_lot::{Condvar, Mutex};
+
+/// Enables multiple threads to synchronize the beginning or end of some computation.
+///
+/// # Examples
+///
+/// ```
+/// use crossbeam_utils::sync::WaitGroup;
+/// use std::thread;
+///
+/// // Create a new wait group.
+/// let wg = WaitGroup::new();
+///
+/// for _ in 0..4 {
+///     // Create another reference to the wait group.
+///     let wg = wg.clone();
+///
+///     thread::spawn(move || {
+///         // Do some work.
+///
+///         // Drop the reference to the wait group.
+///         drop(wg);
+///     });
+/// }
+///
+/// // Block until all threads have finished their work.
+/// wg.wait();
+/// ```
+///
+/// [`Barrier`]: https://doc.rust-lang.org/std/sync/struct.Barrier.html
+pub struct WaitGroup {
+    inner: Arc<Inner>,
+}
+
+/// Inner state of a `WaitGroup`.
+struct Inner {
+    cvar: Condvar,
+    count: Mutex<usize>,
```

Comment:
I guess we can make it `AtomicUsize`.

Reply:
True, but it's not exactly trivial. Maybe we can add the primitive now and optimize it later?